registry/registry_pubkey_update.go

```diff
@@ -0,0 +1,296 @@
+			if err != nil {
+				log.Errorf("Failed to check if namespace already exists: %v", err)
+				return false, nil, errors.Wrap(err, "Server encountered an error checking if namespace already exists")
+			}
+			if exists {
+				// Update the namespace's public key with the latest one when authorized origin provides a new key
+				existingNs, err := getNamespaceByPrefix(prefix)
+				if err != nil {
+					log.Errorf("Failed to get existing namespace to update: %v", err)
+					return false, nil, errors.Wrap(err, "Server encountered an error getting existing namespace to update")
+				}
+
+				// Check the origin is authorized to update (possessing the public key used for prefix initial registration)
+				// Parse all public keys of the sender into a JWKS
+				var clientKeySet jwk.Set
+				if data.AllPubkeys == nil { // backward compatibility - AllPubkeys only exists in the payload in Pelican 7.12 or later
+					clientKeySet, err = jwk.Parse(data.Pubkey)
+				} else {
+					clientKeySet, err = jwk.Parse(data.AllPubkeys)
+				}
+				if err != nil {
+					log.Errorf("Failed to parse in-memory public keys of the client: %v", err)
+					return false, nil, errors.Wrapf(err, "Invalid in-memory public keys format of the client")
+				}
+				// Parse `existingNs.Pubkey` as a JWKS
+				existingKeySet := jwk.NewSet()
+				err = json.Unmarshal([]byte(existingNs.Pubkey), &existingKeySet)
+				if err != nil {
+					log.Errorf("Failed to parse existing namespace public key as JWKS: %v", err)
+					return false, nil, errors.Wrap(err, "Invalid existing namespace public key format")
+				}
+
+				// Check if any key in `clientKeySet` matches a key in `existingKeySet`
+				existingKeysIter := existingKeySet.Keys(ctx)
+				clientKeysIter := clientKeySet.Keys(ctx)
+				matchFound := false
+
+				for existingKeysIter.Next(ctx) {
+					existingKey := existingKeysIter.Pair().Value.(jwk.Key)
+
+					existingKid, ok := existingKey.Get("kid")
+					if !ok {
+						log.Warnf("Skipping registry db existing key without 'kid'")
+						continue
+					}
+
+					for clientKeysIter.Next(ctx) {
+						clientKey := clientKeysIter.Pair().Value.(jwk.Key)
+
+						clientKid, ok := clientKey.Get("kid")
+						if !ok {
+							log.Warnf("Skipping client key without 'kid'")
+							continue
+						}
+
+						if existingKid == clientKid {
+							// Verify the Proof of Possession of client's previous active private key
+							// Get client's previous public key recorded in db
+							var prevRawkey interface{}
+							if err := existingKey.Raw(&prevRawkey); err != nil {
+								return false, nil, errors.Wrap(err, "failed to generate raw pubkey from client's previous pubkey")
+							}
+							// Get client's previous signature from payload
+							var prevKeyVerified bool
+							if data.ClientPrevSignature == "" {
+								prevKeyVerified = true
+							} else {
+								clientPrevSignature, err := hex.DecodeString(data.ClientPrevSignature)
```

Comment:
This portion determining clientPrevSignature isn't dependent on the iteration of the existing keys or client keys, so move it outside of the doubly nested loop to keep the loop as inexpensive as possible.